content/en/blog/_posts/2021-04-15-local-storage-features-go-beta.md

```diff
@@ -0,0 +1,164 @@
+TopoLVM only works on Kubernetes.
+
+PMEM-CSI created its own storage backplane with communication through
+gRPC calls. Securing that communication depends on TLS certificates,
+which made driver deployment more complicated.
+
+The next problem is scheduling. When volumes get created independently
+of pods ("immediate binding"), the CSI driver must pick a node without
+knowing anything about the pod(s) that are going to use it. Topology
+information then forces those pods to run on the node where the volume
+was created. If other resources like RAM or CPU are exhausted there,
+the pod cannot start. This can be avoided by configuring in the
+storage class that volume creation is meant to wait for the first pod
+that uses a volume (`volumeBinding: WaitForFirstConsumer`). In that
+mode, the Kubernetes scheduler tentatively picks a node based on other
+constraints and then the external provisioner is asked to create a
+volume such that it is usable there. If local storage is exhausted,
+the provisioner can ask for another scheduling round. But without
+information about available capacity, the scheduler might always pick
+the same unsuitable node.
+
+Both TopoLVM and PMEM-CSI solved this with scheduler extenders. This
+works, but it is hard to configure when deploying the driver because
+communication between kube-scheduler and the driver is very dependent
+on how the cluster was set up.
+
+A more common use case for local storage are ephemeral volumes that
+get created for a pod and destroyed together with it. The initial API
+for supporting ephemeral volumes with CSI drivers (hence called "*CSI*
+ephemeral volumes") was designed for light-weight volumes where volume
+creation is unlikely to fail. Volume creation happens after pods have
+been permanently scheduled onto a node, in contrast to the the
+traditional provisioning where volume creation is tried before
+scheduling a pod onto a node. CSI drivers must be modified to support
+"CSI ephemeral volumes", which was done for TopoLVM and PMEM-CSI. But
+due to the design of the feature in Kubernetes, pods can get stuck
+permanently if storage capacity runs out on a node. The scheduler
+extenders try to avoid that, but cannot be 100% reliable.
+
+## ... the better...
+
+Starting with external-provisioner v2.1.0, released for Kubernetes
+1.20, provisioning can be handled by external-provisioner instances
+that get deployed together with the CSI driver on each node
+("distributed provisioning"). There is no need anymore to have a
+central component and thus no need for communication between nodes, at
+least not for provisioning.
+
+A scheduler extender still needs some way to find out about capacity
+on each node. When PMEM-CSI switched to distributed provisioning in
+v0.9.0, this was done by querying the metrics data exposed by the
+local driver containers. But even better is to eliminate the need for
+a scheduler extender completely. Storage capacity tracking, promoted
```

Comment:
Is there a page that “Storage capacity tracking” could hyperlink to?

Reply:
Yes. https://kubernetes.io/docs/concepts/storage/storage-capacity/ and https://kubernetes.io/blog/2020/09/01/ephemeral-volumes-with-storage-capacity-tracking/ is what I added as links.